Fix "spack install" when using Python 3.8 on macos

scheibelp opened the pull request:
See: https://github.com/spack/spack/pull/14187

This is the same as #14187 but created as a branch of the main Spack repository to run additional CI checks.

Fixes #14102

scheibelp committed:
pickle may not work on locally-defined functions: extract locally-defined function into a top-level function def

scheibelp committed:
reimplement StageComposite as interface vs. decorator

scheibelp committed:
extract locally-defined function into top-level definition

scheibelp committed:
remove setattr from stagecomposite

scheibelp committed:
reimplement FetchStrategyComposite without class decorator

scheibelp committed:
classes defining __getattr__ should define __setstate__ and __getstate__ to be pickle-able

scheibelp committed:
fdopen was producing a buffered IO object, so just pass along the FD to the child process

scheibelp committed:
resolve circular import (dso_suffix); extract build_process into module-level function

scheibelp committed:
guard running spack command so that it isn't rerun when importing __main__ from subprocess execution

scheibelp committed:
fix infinite recursion with __getattr__ override

scheibelp committed:
fix pkg references for extracted build_process

scheibelp committed:
child process needs to reinitialize repo to make packages import-able

scheibelp committed:
set inheritability of pipes in logger

scheibelp committed:
Revert "set inheritability of pipes in logger"

This reverts commit 31373e74602b00a3754ebc7ad8a0302b32c0dbfd.

scheibelp committed:
merge from develop (inclinding movement of large chunk of affected logic from package.py to installer.py)

scheibelp committed:
flake edits

scheibelp committed:
additional flake edit

scheibelp committed:
merge from develop

scheibelp committed:
add missing paren; use multiprocessing.Connection to properly transmit FD to child process FD table

scheibelp committed:
replace some lambdas; use FDs vs opened file objects when passing args to multiprocessing.Process

scheibelp committed:
fix bad reference (function used to be defined in Package but has moved to Installer)

scheibelp committed:
pass log filename vs. opened file object (which doesn't pickle)

scheibelp committed:
FD's saved with dup in log_output.__enter__ are strangely closed prematurely. Saving them in __init__ prevents this although I don't know why

scheibelp committed:
remove two of the failed closings of Connection._handle (i.e. two different entities attempt to close the same FD)

scheibelp committed:
initial work on replacing process with thread

scheibelp committed:
fix signal handling for thread-based _writer_daemon

scheibelp committed:
fix .close error messages by explicitly closing Connection objects

scheibelp committed:
pass spec dict vs. spec object to _setup_pkg_and_run (this will avoid the requirement to make Spec and all associated objects pickle-able)

scheibelp committed:
most code changes to allow pickling Spec/Package can be undone now that the spec is passed as a dictionary rather than an object

scheibelp committed:
Revert "fix signal handling for thread-based _writer_daemon"

This reverts commit 8a584d238796675fcfa29a0777ca9bdfb8b490d8.

scheibelp committed:
Revert "initial work on replacing process with thread"

This reverts commit 998acd374c7f142290f3258ecfeb393ef833b3e2.

scheibelp committed:
flake edits

scheibelp committed:
update log_output to work with python 2.x and 3.x

scheibelp committed:
Revert "most code changes to allow pickling Spec/Package can be undone now that the spec is passed as a dictionary rather than an object"

This reverts commit 1dfaeac62cf05bd4e985001e6c4cc6b0c6deae45.

scheibelp committed:
Revert "pass spec dict vs. spec object to _setup_pkg_and_run (this will avoid the requirement to make Spec and all associated objects pickle-able)"

This reverts commit 727e70c88a60dd9a9ecbf18f7398979da6dbbc3b.

scheibelp committed:
make sure not to call function if variable is not set

scheibelp committed:
intermediate work on preserving all forms of log file when creating a log_output context manager

scheibelp committed:
don't recv writer_daemon collected output until stopping the writer_daemon loop

scheibelp committed:
remove debug printout

scheibelp committed:
pipe should not be required to terminate logger daemon: closing all instances of write_fd should result in the termination of the read end of the pipe

scheibelp committed:
make sure to close write_fd for Python < 3.8 or if on Linux (where fork is still used)

scheibelp committed:
save stdout/stderr properly when they don't have associated file numbers

scheibelp committed:
properly transmit repos for tests: these can be modified in-memory without changing the config, so logic has been added to properly reconstruct the RepoPath after forking the build process (before deserializing the package)

scheibelp committed:
flake edit

scheibelp committed:
Merge branch 'develop' into bugfix/fork-function

scheibelp committed:
Revert "MacOS nightly builds: use Python 3.7 in CI"

This reverts commit b1133fab22f969f99ea04f10a920485bd5ed6107.

scheibelp commented:
The Jupyter test failure already exists (see: https://github.com/spack/spack/pull/18129)

The `scipy`/`mpl` Mac OS builds fail for a reason initially not related but it appears that the error handling in `log_output.__exit__` needs to be made slightly more robust (EDIT: error handling now addressed as of 8/25):

```
2020-08-20T23:18:36.1761410Z Fatal Python error: init_fs_encoding: failed to get the Python codec of the filesystem encoding
2020-08-20T23:18:36.1761520Z Python runtime state: core initialized
2020-08-20T23:18:36.1761940Z ModuleNotFoundError: No module named 'encodings'
2020-08-20T23:18:36.1762010Z 
2020-08-20T23:18:36.1762110Z Current thread 0x00000001047e0dc0 (most recent call first):
2020-08-20T23:18:36.1762210Z <no Python frame>
2020-08-20T23:18:37.5337140Z ==> Error: EOFError: 
2020-08-20T23:18:37.5337380Z 
2020-08-20T23:18:37.5337480Z /Users/runner/work/spack/spack/lib/spack/spack/build_environment.py:841, in _setup_pkg_and_run:
2020-08-20T23:18:37.5337630Z         838        tb_string = traceback.format_exc()
2020-08-20T23:18:37.5337730Z         839
2020-08-20T23:18:37.5337860Z         840        # build up some context from the offending package so we can
2020-08-20T23:18:37.5337970Z   >>    841        # show that, too.
2020-08-20T23:18:37.5338070Z         842        package_context = get_package_context(tb)
2020-08-20T23:18:37.5338170Z         843
2020-08-20T23:18:37.5338260Z         844        build_log = None
2020-08-20T23:18:37.5338310Z 
2020-08-20T23:18:37.5338370Z 
2020-08-20T23:18:37.5440300Z ==> Warning: Skipping build of py-mpi4py since py-setuptools failed
2020-08-20T23:18:37.5440850Z ==> Error: Terminating after first install failure: EOFError: 
2020-08-20T23:18:37.5441000Z /Users/runner/work/spack/spack/lib/spack/spack/build_environment.py:841, in _setup_pkg_and_run:
2020-08-20T23:18:37.5441120Z         838        tb_string = traceback.format_exc()
2020-08-20T23:18:37.5441220Z         839
2020-08-20T23:18:37.5441480Z         840        # build up some context from the offending package so we can
2020-08-20T23:18:37.5441670Z   >>    841        # show that, too.
2020-08-20T23:18:37.5441780Z         842        package_context = get_package_context(tb)
2020-08-20T23:18:37.5441880Z         843
2020-08-20T23:18:37.5441970Z         844        build_log = None
2020-08-20T23:18:37.5442030Z 
2020-08-20T23:18:37.5442120Z Traceback (most recent call last):
2020-08-20T23:18:37.5442230Z   File "/Users/runner/work/spack/spack/lib/spack/spack/installer.py", line 1606, in build_process
2020-08-20T23:18:37.5442320Z     phase(pkg.spec, pkg.prefix)
2020-08-20T23:18:37.5442430Z   File "/Users/runner/work/spack/spack/lib/spack/spack/package.py", line 109, in phase_wrapper
2020-08-20T23:18:37.5442540Z     phase(spec, prefix)
2020-08-20T23:18:37.5442640Z   File "/Users/runner/work/spack/spack/lib/spack/spack/build_systems/python.py", line 153, in build
2020-08-20T23:18:37.5443290Z     self.setup_py('build', *args)
2020-08-20T23:18:37.5443420Z   File "/Users/runner/work/spack/spack/lib/spack/spack/build_systems/python.py", line 121, in setup_py
2020-08-20T23:18:37.5443890Z     self.python('-s', setup, '--no-user-cfg', *args, **kwargs)
2020-08-20T23:18:37.5444410Z   File "/Users/runner/work/spack/spack/lib/spack/spack/build_systems/python.py", line 115, in python
2020-08-20T23:18:37.5444570Z     inspect.getmodule(self).python(*args, **kwargs)
2020-08-20T23:18:37.5444810Z   File "/Users/runner/work/spack/spack/lib/spack/spack/util/executable.py", line 188, in __call__
2020-08-20T23:18:37.5445290Z     raise ProcessError('Command exited with status %d:' %
2020-08-20T23:18:37.5445410Z spack.util.executable.ProcessError: Command exited with status 120:
2020-08-20T23:18:37.5445940Z     '/Users/runner/work/spack/spack/opt/spack/darwin-catalina-x86_64/apple-clang-11.0.3/python-3.7.8-som75eo56dwhgfl3jzkvu7wrgqkzs7ze/bin/python3.7' '-s' 'setup.py' '--no-user-cfg' 'build'
2020-08-20T23:18:37.5446040Z 
2020-08-20T23:18:37.5446140Z During handling of the above exception, another exception occurred:
2020-08-20T23:18:37.5446420Z 
2020-08-20T23:18:37.5446520Z Traceback (most recent call last):
2020-08-20T23:18:37.5446640Z   File "/Users/runner/work/spack/spack/lib/spack/spack/build_environment.py", line 825, in _setup_pkg_and_run
2020-08-20T23:18:37.5446750Z     return_value = function(pkg, kwargs)
2020-08-20T23:18:37.5446850Z   File "/Users/runner/work/spack/spack/lib/spack/spack/installer.py", line 1606, in build_process
2020-08-20T23:18:37.5446950Z     phase(pkg.spec, pkg.prefix)
2020-08-20T23:18:37.5447050Z   File "/Users/runner/work/spack/spack/lib/spack/llnl/util/tty/log.py", line 571, in __exit__
2020-08-20T23:18:37.5447140Z     self.echo = self.parent_pipe.recv()
2020-08-20T23:18:37.5447260Z   File "/usr/local/Cellar/python@3.8/3.8.5/Frameworks/Python.framework/Versions/3.8/lib/python3.8/multiprocessing/connection.py", line 250, in recv
2020-08-20T23:18:37.5447370Z     buf = self._recv_bytes()
2020-08-20T23:18:37.5447480Z   File "/usr/local/Cellar/python@3.8/3.8.5/Frameworks/Python.framework/Versions/3.8/lib/python3.8/multiprocessing/connection.py", line 414, in _recv_bytes
2020-08-20T23:18:37.5447600Z     buf = self._recv(4)
2020-08-20T23:18:37.5447710Z   File "/usr/local/Cellar/python@3.8/3.8.5/Frameworks/Python.framework/Versions/3.8/lib/python3.8/multiprocessing/connection.py", line 383, in _recv
2020-08-20T23:18:37.5447820Z     raise EOFError
2020-08-20T23:18:37.5447910Z EOFError
```

scheibelp committed:
trigger rebuilds for all packages

scheibelp committed:
Revert "trigger rebuilds for all packages"

This reverts commit afe050e477b3f214fdaf0144bcf8b9dde858eecb.

scheibelp committed:
attempt to trigger rebuilds by changing all package hashes

scheibelp committed:
Python3 pickle can serialize instance methods: no need to import multiprocessing.reduction

scheibelp committed:
Handle context manager clean-up during exception

scheibelp committed:
issues with dup'ed stdout being lost in log_output.__enter__ appear to be transient: I'm going back to saving stdout in __enter__

alalazo commented on .github/workflows/macos_python.yml:
```diff
@@ -25,9 +25,6 @@ jobs:
     runs-on: macos-latest
     steps:
     - uses: actions/checkout@v2
-    - uses: actions/setup-python@v2
-      with:
-        python-version: 3.7
```
Using this action is recommended by Github, see [here](https://docs.github.com/en/actions/language-and-framework-guides/using-python-with-github-actions#specifying-a-python-version). Maybe we can bump the Python version instead of relying on system Python?

scheibelp committed:
change function/class names to clarify; add docstring

scheibelp committed:
Revert "Revert "MacOS nightly builds: use Python 3.7 in CI""

This reverts commit 3a15333649386a62fdacf2b7b56fe9a960aa8df9.

scheibelp committed:
specify that Mac OS tests should use 3.8 instead of 3.7

scheibelp committed:
restore old method of saving stdout/stderr in __enter__ since I'm no longer encountering issues where the FD is closed after __enter__

scheibelp committed:
remove unused attribute

adamjstewart commented:
Can you also bump Python versions in `macos_unit_tests.yaml`? Many tests are failing for me on this branch when I run with Python 3.8 on macOS.

adamjstewart committed:
Run all CI tests with Python 3.8

adamjstewart reviewed (changes requested):
As discussed over Slack, the unit tests currently fail with Python 3.8 on macOS:
```console
$ spack test lib/spack/spack/test/install.py
...
Process Process-3:
Traceback (most recent call last):
  File "/Users/Adam/spack/opt/spack/darwin-catalina-x86_64/apple-clang-11.0.3/python-3.8.5-xoylgs5c4g3l7duxlff3gubiv5c7arqa/lib/python3.8/multiprocessing/process.py", line 315, in _bootstrap
    self.run()
  File "/Users/Adam/spack/opt/spack/darwin-catalina-x86_64/apple-clang-11.0.3/python-3.8.5-xoylgs5c4g3l7duxlff3gubiv5c7arqa/lib/python3.8/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/Users/Adam/spack/lib/spack/spack/build_environment.py", line 816, in _setup_pkg_and_run
    pkg = serialized_pkg.restore()
  File "/Users/Adam/spack/lib/spack/spack/build_environment.py", line 879, in restore
    return pickle.load(self.serialized_pkg)
  File "/Users/Adam/spack/lib/spack/spack/test/install.py", line 86, in __getattr__
    return getattr(self.wrapped_stage, attr)
  File "/Users/Adam/spack/lib/spack/spack/test/install.py", line 86, in __getattr__
    return getattr(self.wrapped_stage, attr)
  File "/Users/Adam/spack/lib/spack/spack/test/install.py", line 86, in __getattr__
    return getattr(self.wrapped_stage, attr)
  [Previous line repeated 989 more times]
RecursionError: maximum recursion depth exceeded
Exception ignored in: <function _ConnectionBase.__del__ at 0x10fae8e50>
Traceback (most recent call last):
  File "/Users/Adam/spack/opt/spack/darwin-catalina-x86_64/apple-clang-11.0.3/python-3.8.5-xoylgs5c4g3l7duxlff3gubiv5c7arqa/lib/python3.8/multiprocessing/connection.py", line 132, in __del__
  File "/Users/Adam/spack/opt/spack/darwin-catalina-x86_64/apple-clang-11.0.3/python-3.8.5-xoylgs5c4g3l7duxlff3gubiv5c7arqa/lib/python3.8/multiprocessing/connection.py", line 361, in _close
OSError: [Errno 9] Bad file descriptor
```

scheibelp committed:
intermediate work on transferring global state as well as test monkeypatches to classes/modules; Package instances can creep into the references in unexpected ways

scheibelp committed:
make MockStage pickle-able

scheibelp committed:
forgot file needed by prior commit

scheibelp committed:
serialize transmitted monkeypatches; also use monkeypatch in a few places which were 'hand rolling' an alternative to monkeypatch

scheibelp committed:
merge from develop (needs additional work to resolve hangs)

scheibelp committed:
remove other all uses of fork_context and use multiprocessing (which uses the 'spawn' method vs. forking)

scheibelp committed:
flake edits

scheibelp committed:
add new license

scheibelp committed:
fix tests based on new list_url searching logic

scheibelp committed:
now that monkeypatch patches are only deserialized after repo construction, Package.fetch_remote_versions can be restored

scheibelp committed:
fix pickling issues with most lock tests when using spawn process start method

scheibelp committed:
update more lock tests to replace locally-defined functions with classes (to make them pickle-able)

scheibelp committed:
flake edit

scheibelp committed:
allow module file overwrite for test modules

scheibelp committed:
add process object which handles transmission of test state for spack tests that use the db etc. outside of installation

scheibelp committed:
flake edit

scheibelp committed:
it is no longer necessary for packages to explicitly call record_monkeypatch; a session-scoped fixture now updates monkeypatch to record instances of calls to setattr

scheibelp committed:
flake edit

scheibelp committed:
forgot module reference

scheibelp committed:
resolve unpickling error for Singleton objects (request for attribute that is not defined yet leads to infinite recursion)

scheibelp committed:
extract another locally-defined function to make it pickle-able in tests

scheibelp committed:
skip no_termios v-toggle test for python 3.8 on Mac (TODO: investigate why this test doesnt pass)

scheibelp committed:
proper handling of file closure across 3.8 and versions < 3.8

scheibelp committed:
improve explanatory comment

scheibelp committed:
merge from develop (conflict in installer.py)

scheibelp committed:
extract locally defined function for pickling

scheibelp committed:
fix 'file descriptor out of range in select' (likely caused by failing to close files in python < 3.8)

scheibelp committed:
merge from develop (resolve conflicts introduced by since-reverted PR)

scheibelp committed:
run unit tests for Linux/Mac with Python 3.9 (in addition to 3.8)

becker33 commented on lib/spack/llnl/util/tty/log.py:
```diff
@@ -519,14 +583,14 @@ def __exit__(self, exc_type, exc_val, exc_tb):
+            # _writer_daemon. An exception will have already been generated.
+            pass
 
-        # join the daemon process. The daemon will quit automatically
-        # when the write pipe is closed; we just wait for it here.
```
Why is this comment removed? Is the behavior not still accurate?

becker33 commented on lib/spack/llnl/util/tty/log.py:
```diff
@@ -593,25 +658,37 @@ def _writer_daemon(stdin, read_fd, write_fd, echo, log_file, control_pipe):
-        read_fd (int): pipe for reading from parent's redirected stdout
-        write_fd (int): parent's end of the pipe will write to (will be
-            immediately closed by the writer daemon)
+        read (int): pipe for reading from parent's redirected stdout
```
This docstring no longer matches the function signature

becker33 commented on lib/spack/llnl/util/tty/log.py:
```diff
@@ -593,25 +658,37 @@ def _writer_daemon(stdin, read_fd, write_fd, echo, log_file, control_pipe):
+    in_pipe = os.fdopen(read_wrapper._handle, 'r', 1)
+
+    if stdin_wrapper:
+        stdin = os.fdopen(stdin_wrapper._handle)
```
If we need to access the `_handle` attribute on FileWrappers, then we need to make it a public attribute of the interface.

It also looks like we have inconsistent naming between wrappers and `MultiProcessFd`s. It's the `MultiProcessFd` class that has a `_handle` attribute, but this variables are named `*_wrapper`. I think we need to work to clarify this (either in code or comments) to make it more maintainable.

becker33 commented on lib/spack/llnl/util/tty/log.py:
```diff
@@ -593,25 +658,37 @@ def _writer_daemon(stdin, read_fd, write_fd, echo, log_file, control_pipe):
     istreams = [in_pipe, stdin] if stdin else [in_pipe]
     force_echo = False      # parent can force echo for certain output
 
+    log_file = log_file_wrapper.unwrap()
```
This seems fragile. We're taking a class/method that can return a stream or a filename, and then we're assuming it's a stream and writing to it.

Is there any reason we can't either:
1. Have an `open` method (or option to `unwrap`) which ensures the returned object is of appropriate type (to write to iff open=True, to later open if open=False or something of the sort)
2. Use separate classes to wrap streams vs filenames

scheibelp replied:
> This seems fragile. We're taking a class/method that can return a stream or a filename, and then we're assuming it's a stream and writing to it.

`FileWrapper.unwrap` is always supposed to return a file-like object (either a file, or a `StringIO`). It should never return a string.

EDIT: it might contain a string internally, but the process of `unwrap`ping it should produce a `File`

becker33 replied:
What if it's constructed with a string, and `open` is set to `False`. Wouldn't that just return a string?

scheibelp replied:
It would, but it shouldn't be called with that mix, so I should guard against that in `FileWrapper.__init__`

scheibelp replied:
I have moved relevant logic into the `FileWrapper` to make it determine this state itself (and thereby avoid any issue where it can be improperly initialized).

becker33 commented on lib/spack/spack/build_environment.py:
```diff
@@ -803,7 +804,94 @@ def modifications_from_dependencies(spec, context):
+            input_multiprocess_fd.close()
+
+
+class TransmitPackage(object):
```
Can we rename this to `ContextTransmitter`? For two reasons

1. It transmits more than just the package
2. I think it's good practice for class names to be nouns, not verbs

scheibelp replied:
I can come up with a better name, but I think it's worth noting that outside of a test context, the only thing that needs to be transmitted is the package: all the other serialization (i.e. that goes on inside `TransmitTestState`) is only required for tests (because they modify configs typically read from disk in memory and expect the subprocess to reflect that).

becker33 replied:
But that might not always be true, even if it is right now. I'd rather have a general method, that we can then "dogfood" by using it for our tests, than have a special case for passing tests.

becker33 commented on lib/spack/spack/build_environment.py:
```diff
@@ -831,63 +919,21 @@ def child_fun():
-
-        finally:
-            child_pipe.close()
+    serialized_pkg = TransmitPackage(pkg)
```
Similarly this should be renamed to make clear it's more than just the package.

becker33 commented on lib/spack/spack/installer.py:
```diff
@@ -1035,15 +1035,13 @@ def _install_task(self, task, **kwargs):
-        verbose = kwargs.get('verbose', False)
+
+        build_kwargs = kwargs.copy()
+        build_kwargs['dirty'] = kwargs.get('dirty', False)
+        build_kwargs['fake'] = kwargs.get('fake', False)
```
We can't we handle the defaulting of these methods later without a separate set of args?

scheibelp replied:
I don't actually know whether this is required, but I like copying the state because then I don't have to worry whether it's important or not.

becker33 replied:
Even if you copy them, you shouldn't have to set `dirty` and `fake` if we properly interpret the defaults for them on the other side.

becker33 commented on lib/spack/spack/installer.py:
```diff
@@ -1621,6 +1515,116 @@ def spec(self):
+
+    This function's return value is returned to the parent process.
+    """
+    kwargs = build_kwargs
+
+    keep_stage = kwargs.get('keep_stage', False)
+    install_source = kwargs.get('install_source', False)
+    skip_patch = kwargs.get('skip_patch', False)
+    verbose = kwargs.get('verbose', False)
+    fake = kwargs.get('fake', False)
+
```
Is anything after this point in this method changed? It looks like it was just a basic "dedent" operation.

scheibelp replied:
It is almost exactly the same. This change was made because pickling doesn't want to deal with functions defined inside of functions. The following sections are different:

```
     pkg_id = package_id(pkg)
     tty.debug('{0} Building {1} [{2}]'
               .format(pre, pkg_id, pkg.build_system_class))

     # get verbosity from do_install() parameter or saved value
     echo = verbose
     if spack.package.PackageBase._verbose is not None:
         echo = spack.package.PackageBase._verbose

     pkg.stage.keep = keep_stage

     # parent process already has a prefix write lock
     with pkg.stage:
```

and

```
 pid = '{0}: '.format(pkg.pid) if tty.show_pid() else ''
 pre = '{0}{1}:'.format(pid, pkg.name)
 pkg_id = package_id(pkg)

 tty.debug('{0} Building {1} [{2}]'
           .format(pre, pkg_id, pkg.build_system_class))

 # get verbosity from do_install() parameter or saved value
 echo = verbose
 if spack.package.PackageBase._verbose is not None:
     echo = spack.package.PackageBase._verbose

 pkg.stage.keep = keep_stage
 with pkg.stage:
```

becker33 commented on lib/spack/spack/spec.py:
```diff
@@ -1403,6 +1408,7 @@ def _spec_hash(self, hash):
         # this when we move to using package hashing on all specs.
         node_dict = self.to_node_dict(hash=hash)
         yaml_text = syaml.dump(node_dict, default_flow_style=True)
+        yaml_text += 'change-all-hashes'
```
Is this left over from debugging?

scheibelp replied:
Yes: this was an attempt to get CI to rebuild all packages under consideration (i.e. it would see a new hash and assume the package has changed).

becker33 replied:
Ok, then I'm assuming we should remove it before merging.

becker33 commented on lib/spack/spack/spec.py:
```diff
@@ -174,6 +174,10 @@
 
 
+# Platform-specific library suffix.
+dso_suffix = 'dylib' if sys.platform == 'darwin' else 'so'
```
This seems like the wrong module for this information

becker33 commented on lib/spack/spack/test/conftest.py:
```diff
@@ -744,34 +817,7 @@ def module_configuration(monkeypatch, request):
         spack.paths.test_path, 'data', 'modules', writer_key
     )
 
-    def _impl(filename):
```
Are methods defined in a method bad for pickle? I know classes defined at non-module scope are a problem, but I thought you could pickle functions defined at any scope.

scheibelp replied:
> Are methods defined in a method bad for pickle?

Yes

becker33 replied:
We should probably add something to that effect to the developer guide. I don't think it should be a problem to ban nested methods and nested functions from our style guide.

becker33 commented on lib/spack/spack/test/database.py:
```diff
@@ -517,14 +516,21 @@ def test_026_reindex_after_deprecate(mutable_database):
-    p = fork_context.Process(target=read_and_modify, args=())
+
+def test_030_db_sanity_from_another_process(mutable_database):
+    action = ReadModify()
```
If we're going to name this as a verb, we might as well use `__call__` and make it a true callable.

becker33 commented on lib/spack/spack/test_state.py:
```diff
@@ -0,0 +1,114 @@
+        fn()
+
+    def create(self):
+        test_state = TransmitTestState()
+        return multiprocessing.Process(
+            target=self._restore_and_run,
+            args=(self.fn, test_state))
+
+
+class TransmitTestState(object):
```
I'd like to change the names here for clarity as well. `TransmitTestSTate()` makes me think something has been done, not merely initialized.

becker33 commented on lib/spack/spack/test_state.py:
```diff
@@ -0,0 +1,114 @@
+        fn()
+
+    def create(self):
+        test_state = TransmitTestState()
+        return multiprocessing.Process(
+            target=self._restore_and_run,
+            args=(self.fn, test_state))
+
+
+class TransmitTestState(object):
```
I think this should be refactored to be more general, and combined with the package serialization code.

Could we have a `class SpackContextSerializer`, and then different types of objects could be registered for serialization? The package will be one of those, and monkeypatch patches would be another, etc. I think that will make the code a little more stable and easier to maintian/extend.

Is there something I'm missing that would make that difficult/impossible?

scheibelp replied:
> Could we have a class SpackContextSerializer, and then different types of objects could be registered for serialization?

If I understand, this class would have a `register` method and I could add all of the objects of interest. The issue with that is that most of the objects passed to `TransmitTestState` are serialized in different ways:

* The monkeypatches from the test must be explicitly serialized because the test repositories must be available when they are unpickled (so first the `Repo` must be reconstructed, then the patches deserialized).
* The `Repo` object is serialized without `pickle`: I just store the directories. In general if an object is not pickle-able but has a straightforward serialization method, I went with that.

(`config` and `architecture.platform` can be stored as attributes on the object: they serialize/deserialize without problems and order doesn't matter for them)

The second point isn't critical - I think `Repo` could be made pickle-able (without that each object would have to be registered along with a serialization method) - but the first means that a generic `register` method would also need a priority. In short, the generality required to support the specific serialization use cases for the Spack installation would make it fairly complex, and I don't know of other cases where this would be useful (e.g. typically you don't need a special object to serialize/deserialize: you can just pass objects as additional args to `multiprocessing.Process`).

becker33 replied:
If the registered objects were held in a list, and deserialized in order, would that be sufficient?

What I really don't like is that the `PackageTransmitter` is responsible for passing the test state in addition to the package itself. That seems like it breaks separation of concerns.

becker33 commented on lib/spack/spack/test_state.py:
```diff
@@ -0,0 +1,114 @@
+    _serialize = sys.version_info >= (3, 8) and sys.platform == 'darwin'
+
+    def __init__(self):
+        if TransmitTestState._serialize:
```
Can't this use `self._serialize`?

scheibelp replied:
It could but I prefer to access class attributes via the class: I think it's useful to remind readers that an attribute is a class attribute whenever possible (since the semantics of interacting with it are different vs. an instance attribute).

becker33 replied:
Oh interesting. I disagree, because I think class inheritance is something readers understand easily, and using the instance to access the class makes obvious that it is from its own class -- when I read this code, the first thing I thought was that this was a method from a class that the object did not belong to.

scheibelp replied:
For now this has been sidestepped by promoting `_serialize` to a module level variable (not directly because of this conversation but rather because we spoke elsewhere about aggregating the package/other-state serialization).

> using the instance to access the class makes obvious that it is from its own class

That's true, but I think a user has more to lose when they mistake a class variable for an instance variable (more likely to set it) vs. when they think a class variable belongs to another class (generally I assume folks will be more careful with setting class variables, regardless of whether they belong to the class which self is an instance of or not.

tldahlgren commented on lib/spack/llnl/util/lang.py:
```diff
@@ -568,6 +568,8 @@ def instance(self):
         return self._instance
 
     def __getattr__(self, name):
+        if name == '_instance' or name == 'instance':
```
Why not `name in ['_instance', 'instance']` ?

becker33 commented on lib/spack/spack/repo.py:
```diff
@@ -1164,14 +1164,14 @@ def create_or_construct(path, namespace=None):
 
 
-def _path():
+def _path(repo_dirs=None):
```
Is this just to make it easier to patch this for testing?

scheibelp replied:
The change is for tests, but I would say that it is necessary, since I need to transmit the in-memory state of the test process to the subprocess, and without this change the `RepoPath` is only constructed from disk (e.g. bypassing the mock repo established for the test).

scheibelp committed:
resolving review comments - part 1

scheibelp committed:
review comments: resolve wrapper/multiprocess_fd naming inconsistency; expose private Connection._handle via MultiprocessFd interface and remove separate multiprocess_fd method

scheibelp committed:
move logic into FileWrapper in attempt to reduce confusion about its possible states

scheibelp committed:
remove unnecessary copy of kwargs

scheibelp committed:
add __call__ method to action-based object to get more idiomatic usage

scheibelp committed:
fix typos; kwargs defaults; rename TransmitPackage to PackageInstallContext

scheibelp committed:
rename some package/state serialization handlers and add clarifying comments

scheibelp committed:
undo movement of dso_suffix definition

scheibelp committed:
merge from develop (conflicts in stage/github macos unit tests)

scheibelp committed:
flake edits

scheibelp commented:
I have resolved most comments although unfortunately had a bit of trouble with the biggest ones. Ones that I think I resolved in an uncontroversial manner I have marked as resolved. The remaining ones I have left open and added further discussion here.

> I think this should be refactored to be more general, and combined with the package serialization code.

 I realized the latter part of this doesn’t fit the current needs of the packages:

- package tests doing installations need to serialize (the package) plus (the test monkey patches, the repo, the config, and the platform
- Some tests don’t involve package installations but need to serialize other monkey patches to a subprocess (this is only needed once) - see SpackTestProcess

So it is necessary to sever these. Furthermore, `build_environment.fork` needs to specifically retrieve the Package object, since that’s what it does the install with, so I’m having trouble coming up with a generic serialization interface which accommodates that specific requirement.

> What I really don't like is that the PackageTransmitter is responsible for passing the test state in addition to the package itself. That seems like it breaks separation of concerns.

I can either embed the transmission of test state in this object or bring it outside. In the latter case, I’m not sure what other object should know about this. For example I don’t think build_environment should have to explicitly understand that it needs to serialize the test state and pass it to the installation subprocess.

>> all the other serialization (i.e. that goes on inside TransmitTestState) is only required for tests (because they modify configs typically read from disk in memory and expect the subprocess to reflect that).

> I'd rather have a general method, that we can then "dogfood" by using it for our tests, than have a special case for passing tests.

(https://github.com/spack/spack/pull/18205#discussion_r505786823)

I agree that the test behavior should match normal behavior where possible; to be clear, that is how this PR has behaved. I have clarified it with comments/renaming of classes since the review.

scheibelp commented on lib/spack/spack/test_state.py:
```diff
@@ -0,0 +1,139 @@
+            args=(self.fn, test_state))
+
+
+class PackageInstallContext(object):
```
I have renamed `TransmitPackage` to `PackageInstallContext`. I also renamed `TransmitTestState` to `TestState` (plus added a docstring for clarification) and `TransmitPatches` to `TestPatches`

scheibelp commented:
@becker33 @tldahlgren I have addressed most comments and have a few questions on how to proceed at https://github.com/spack/spack/pull/18205#issuecomment-715199088

scheibelp committed:
merge from develop (minor conflict in installer.py)

becker33 reviewed (commented):
@scheibelp This mostly looks good to me, with the exception that we should use the spawn method at all times instead of just on macos python3.8. Is that a simple addition to make or does it need to be a separate PR?

scheibelp committed:
add debugging output for failing sbang test (only fails in CI not locally)

scheibelp committed:
remove pdb drop-in

scheibelp committed:
sbang errors were a false issue - a different failing test was creating those failures. fix pickling of locally-defined method.

scheibelp committed:
extract another function for pickling

scheibelp committed:
check if there's an issue with accessing store.layout.root at module import time vs. at test time

scheibelp committed:
update tests for shebang line length on Mac OS

adamjstewart commented:
> we should use the spawn method at all times instead of just on macos python3.8

Why is this the case?

scheibelp commented:
Close/reopen to re-initialize tests

scheibelp commented:
Close/reopen to restart tests

scheibelp committed:
start _writer_daemon without by using the environment as it was before applying updates related to installing the package (running the _writer_daemon process requires the environment that was used to start Spack itself)

becker33 commented:
@adamjstewart because there's a refactor of `spack.util.executable.Executable` that @tgamblin, @scheibelp, and I have been discussing that would allow us to simplify the `spack test` PR that @tldahlgren and I have been working on, but it is infeasible as long as we use `fork` because of how `fork` interacts with `multiprocess.ThreadPool`, which we use an `Executable` inside of for the compiler detection.

tgamblin commented on lib/spack/llnl/util/lang.py:
```diff
@@ -568,6 +564,8 @@ def instance(self):
 
     def __getattr__(self, name):
+        if name in ['_instance', 'instance']:
+            raise AttributeError()
```
This could use a comment saying why we do it -- I get it but I think it's easy to sit and think about it if you don't understand our singletons that well.

tgamblin commented on lib/spack/llnl/util/tty/log.py:
```diff
@@ -288,6 +292,104 @@ def _file_descriptors_work(*streams):
+
+
+@contextmanager
+def replace_environment(env):
```
needs a docstring

tgamblin commented on lib/spack/llnl/util/tty/log.py:
```diff
@@ -592,26 +695,38 @@ def _writer_daemon(stdin, read_fd, write_fd, echo, log_file, control_pipe):
         echo (bool): initial echo setting -- controlled by user and
             preserved across multiple writer daemons
-        log_file (file-like): file to log all output
+        log_file_wrapper (file-like): file to log all output
```
is it still a `file-like` or should it be a `FileWrapper` now?

tgamblin commented on lib/spack/llnl/util/tty/log.py:
```diff
@@ -592,26 +695,38 @@ def _writer_daemon(stdin, read_fd, write_fd, echo, log_file, control_pipe):
+    # If this process was forked, then it will inherit file descriptors from
+    # the parent process. This process depends on closing all instances of
+    # write_fd to terminate the reading loop, so we close the file descriptor
+    # here. Forking is the process spawning method for Python < 3
```
Is this true on Windows?  Can Python 2 be made to spawn as well so that we do not have a separate code path for 2 and 3 here?

scheibelp replied:
This comment looks like it might be from a while ago, but the part about Windows not using fork is definitely true.

scheibelp committed:
undo changing all hashes (for rebuilding in CI)

scheibelp committed:
rename 'test_state' module to 'subprocess_context' and add module description

scheibelp commented:
@becker33 @tgamblin discussed changes are now complete:

* Reverted changes that use spawn start method everywhere (for now)
* Renamed `test_state` module to `subprocess_context` (and added docstring)
* Added logic to revert environment changes when starting `_writer_daemon`: https://github.com/spack/spack/pull/18205/commits/6bbb7debfe69438318aaeafa608f4c5c6ee6a606

@tldahlgren https://github.com/spack/spack/pull/18205#discussion_r504998488 is now addressed.

tgamblin commented on lib/spack/spack/build_environment.py:
```diff
@@ -803,7 +805,66 @@ def modifications_from_dependencies(spec, context):
+
+
+def fork(pkg, function, kwargs):
     """Fork a child process to do part of a spack build.
```
`fork()` should probably have a new name, but at the very least the comment here should indicate that we don't always actually fork -- can you update this to describe what's done and on what systems/python versions?

tgamblin commented on lib/spack/spack/hooks/sbang.py:
```diff
@@ -17,9 +17,13 @@
-shebang_limit = 127
+#: Character limit for shebang line: 127 for Linux; 511 for Mac. Mac
+#: generates longer temporary file paths, which can lead to failures when
+#: generating sbang scripts for tests.
```
This should mention that the numbers here are actually the OS limits (see https://www.in-ulm.de/~mascheck/various/shebang/).  I'm not sure that the point about temp file paths is relevant in the `sbang` hook module.

tgamblin reviewed (changes requested):
@scheibelp: I have a few very minor change requests.

adamjstewart commented on .github/workflows/macos_unit_tests.yaml:
```diff
@@ -12,13 +12,16 @@ on:
     runs-on: macos-latest
+    strategy:
+      matrix:
+        python-version: [3.8, 3.9]
```
Is there any reason to test both? This will slow down unit tests as there are always a shortage of macOS runners.

scheibelp replied:
I think some Mac OS distributions are using 3.8 now but not 3.9. I think if we only test one it would be ideal to test the one that is more-widely used.

adamjstewart replied:
In that case I would only test 3.8, that's the version that ships on the latest macOS.

adamjstewart commented on lib/spack/spack/test/llnl/util/tty/log.py:
```diff
@@ -396,6 +396,11 @@ def mock_shell_v_v_no_termios(proc, ctl, **kwargs):
 def test_foreground_background_output(
         test_fn, capfd, termios_on_or_off, tmpdir):
     """Tests hitting 'v' toggles output, and that force_echo works."""
+    if (sys.version_info >= (3, 8) and sys.platform == 'darwin'
+        and termios_on_or_off == no_termios):
+
+        return
```
It might be better to do this as a `pytest.mark.skipif` instead. Can you have multiple?

scheibelp replied:
You can have multiple, but I don't know how to do it based on the parameterization (i.e. I can check for the python version and platform, but I'm not sure how I would access `termios_on_or_off`).

scheibelp committed:
address most review comments

scheibelp committed:
rename 'fork' to 'start_child_process'

scheibelp committed:
rename 'start_child_process' to 'start_build_process'

scheibelp commented:
@tgamblin all review comments are now addressed.

I left https://github.com/spack/spack/pull/18205#discussion_r521628073 unresolved because I only handled the part about changing the comment to clarify how things work on Windows. I did not do

> Can Python 2 be made to spawn as well so that we do not have a separate code path for 2 and 3 here?

because I think we already spoke about skipping this for now (I got the impression this was an old review comment because it was out of date before I made any edits).

scheibelp commented:
@adamjstewart https://github.com/spack/spack/pull/18205#pullrequestreview-477090780 is now addressed

becker33 commented:
Closing and reopening to kick off the tests that's wedged.
